File: arrow-libs/fx/arrow-benchmarks-fx/arrow-kio-benchmarks/src/main/kotlin/kio/AttemptRaisedError.kt
package arrow.benchmarks.effects.kio

import it.msec.kio.Task
import it.msec.kio.attempt
import it.msec.kio.effect
import it.msec.kio.just
import it.msec.kio.tryRecover

object AttemptRaisedError {

  private val dummy = object : RuntimeException("dummy") {
    override fun fillInStackTrace(): Throwable =
      this
  }

  private fun loopNotHappy(size: Int, i: Int): Task<Int> =
    if (i < size) {
      effect { throw dummy }
        .attempt()
        .tryRecover { loopNotHappy(size, i + 1) }
    } else
      just(1)

  fun attemptRaisedError(size: Int) =
    it.msec.kio.runtime.Runtime.unsafeRunSync(loopNotHappy(size, 0))
}

File: arrow-libs/fx/arrow-benchmarks-fx/arrow-scala-benchmarks/src/main/scala/cats/AttemptRaisedError.scala
package arrow.benchmarks.effects.scala.cats

import cats.effect.IO

object dummy extends RuntimeException {
  override def fillInStackTrace(): Throwable = this
}

object AttemptRaisedError {

  def ioLoopNotHappy(size: Int, i: Int): IO[Int] =
    if (i < size) {
      IO {
        throw dummy
      }.attempt.flatMap { it =>
        it.fold(_ => ioLoopNotHappy(size, i + 1), IO.pure)
      }
    } else IO.pure(1)

}
